### ISSUE TITLE
Path and Name confusion

### ISSUE THREAD

**Issue by SeanFeldman** [Azure/azure-service-bus-dotnet#67]
This is more of a question/discussion issue.
With the old client, there was somewhat of a confusion around Path and Name.

#### Queues

`this/is/a/queue`
[path](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.queuedescription#Microsoft_ServiceBus_Messaging_QueueDescription_Path): `this/is/a/queue` 

#### Topics
`a/b/c`
[path](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.topicdescription#Microsoft_ServiceBus_Messaging_TopicDescription_Path): `a/b/c`

#### Subscription
`some/topic/subscriptions/sub`
[topic path](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.subscriptiondescription#Microsoft_ServiceBus_Messaging_SubscriptionDescription_TopicPath): `some/topic`
[name](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.subscriptiondescription#Microsoft_ServiceBus_Messaging_SubscriptionDescription_Name): `sub`

While queues and topics share a common `Path` concept, they are not consistent with subscriptions. Would it helpful to have a `FullPath` abstraction that would be available for all 3 entity types?

Queues: `FullPath` == `Path`
Topics: `FullPath` == `Path`
Subscriptions: `FullPath` == `TopicPath`/subscriptions/`Name`

**Comment by jtaubensee** [comment]
@SeanFeldman - What you are saying makes sense. Would it be more clear if we just provided the name and got rid of the path all together? Is there any scenario you can think of where the path is particularly important?

**Comment by SeanFeldman** [comment]
Path makes more sense to me tbh as it can have segments (separated by `/`).
The outlier is subscriptions. Since they have a segment that breaks that pattern therefore there's a name that makes sense. I'd like to keep a name for subscriptions since at the application level you refer to a subscription by its name, not path. Hence the suggest to have a `FullPath`. What does the team say?

**Comment by abombss** [comment]
Not just subscriptions but $deadletter as well. Would love to see the full path concept instead of name. Personally I hate having to deal with queues, topics, and subscriptions differently. They are just senders or receivers in my book with an address.

**Comment by jtaubensee** [comment]
Hey Guys - Just doing some cleanup here. 

@SeanFeldman - I think we are doing what you are asking for. In the subscription client, we provide the path as `${topicpath}/Subscriptions/{subscriptionName}`, and in the queue/topic client we provide the path as simply the name. In your case `some/topic` is the name.

@abombss - Have you used the [`EntityNameHelper`](https://github.com/Azure/azure-service-bus-dotnet/blob/dev/src/Microsoft.Azure.ServiceBus/Primitives/EntityNameHelper.cs) at all? It will help you with formatting deadletter, subscription paths, etc.

I'm going to close this issue for now, but feel free to comment back in case I missed anything.

**Comment by SeanFeldman** [comment]
@jtaubensee does that mean all entities now will have an equivalent of my suggested `FullPath` property that allows to use `MessageSender`/`MessageReceiver`?

**Comment by jtaubensee** [comment]
[`QueueClient.Path`](https://github.com/Azure/azure-service-bus-dotnet/blob/dev/src/Microsoft.Azure.ServiceBus/QueueClient.cs#L77):
```csharp
public string QueueName { get; }
public string Path => this.QueueName;
```

[`TopicClient`](https://github.com/Azure/azure-service-bus-dotnet/blob/dev/src/Microsoft.Azure.ServiceBus/TopicClient.cs#L63)
```csharp
public string TopicName { get; }
```

[`SubscriptionClient`](https://github.com/Azure/azure-service-bus-dotnet/blob/dev/src/Microsoft.Azure.ServiceBus/SubscriptionClient.cs#L78)
```csharp
public string TopicPath { get; }
public string Path { get; }
public string SubscriptionName { get; }

// set in constructor
this.Path = EntityNameHelper.FormatSubscriptionPath(this.TopicPath, this.SubscriptionName);
```

It looks like we are missing the `Path` in the `TopicClient`, so I will add it. It should look like this:
```csharp
public string Path => this.TopicName;
```

**Comment by jtaubensee** [comment]
Closing based on #149

**Comment by SeanFeldman** [comment]
Reopening as there still seems to be some confusion around `path` vs `name`.
For example, `QueueDescription` works with [`Path`](https://github.com/Azure/azure-service-bus-dotnet/blob/edfa095a1eb2102d6ca7d45f030c9df75c489ad2/src/Microsoft.Azure.ServiceBus/Management/QueueDescription.cs#L27). The management client, treats it as a `name` ([`IManagementClient.GetQueueAsync(string queueName)`](https://github.com/Azure/azure-service-bus-dotnet/blob/edfa095a1eb2102d6ca7d45f030c9df75c489ad2/src/Microsoft.Azure.ServiceBus/Management/ManagementClient.cs#L123)).

**Comment by nemakam** [comment]
Will be resolved in #481.
